### PR TITLE
i/apparmor: support for home.d tunables from /etc/

### PR DESCRIPTION
**Pull request by Meulengracht**
The release of snapd 2.60.2 surfaced a bug in the new vendored apparmor integration code. For users with customized /home dirs there is a regression that prevents snaps from working. This is reported in https://forum.snapcraft.io/t/upgrade-snapd-from-2-59-5-to-2-60-2-gives-apparmor-denied-on-app-launch-for-every-app/36492 and also in various LP bugreports.

The issue is that the vendored apparmor does not seem to use the system /etc because of the "--base" keyword we use in the apparmor_parser code. This branch fixes this by ensuring that at least /etc/apparmor.d/tunabables/home.d is included.

More work is needed to include the other tunables, I suspect the easiest fix for this is to add a new option to apparmor_parser to have a commandline option for a tunables search path but this requires more work and short term this branch should help the affected users.

### CODE REVIEW

File: sandbox/apparmor/apparmor.go [hunk 1]
```diff
@@ -460,6 +460,7 @@ func AppArmorParser() (cmd *exec.Cmd, internal bool, err error) {
 			args := []string{
 				"--config-file", filepath.Join(prefix, "/apparmor/parser.conf"),
 				"--base", filepath.Join(prefix, "/apparmor.d"),
+				"--include", "/etc/apparmor.d/tunables/home.d",
```
Comment (Meulengracht) [inline sandbox/apparmor/apparmor.go:463]: This is needed to correctly support relative #include statements from files inside `home.d` here

File: tests/main/snapd-homedirs-vendored/task.yaml [hunk 2]
```diff
@@ -0,0 +1,78 @@
+summary: Test support for non-standard home directory paths
+
+systems:
```
Comment (Meulengracht) [inline tests/main/snapd-homedirs-vendored/task.yaml:3]: this test can only run on ubuntu 20.04 right now

File: interfaces/apparmor/backend.go [hunk 3]
```diff
@@ -829,6 +842,14 @@ func (b *Backend) addContent(securityTag string, snapInfo *snap.Info, cmdName st
 				return `#include if exists "/var/lib/snapd/apparmor/snap-tuning"`
 			}
 			return ""
+		// XXX: Remove this when we have a better solution to including the system
+		// tunables. See snapConfineFromSnapProfile() for a more detailed explanation.
+		case "###INCLUDE_SYSTEM_TUNABLES_HOME_D_WITH_VENDORED_APPARMOR###":
+			features, _ := parserFeatures()
+			if strutil.ListContains(features, "snapd-internal") {
+				return `#include "/etc/apparmor.d/tunables/home.d"`
```
Comment (alexmurray) [inline interfaces/apparmor/backend.go:850]: same here - this should probably be `#include if exists "/etc/apparmor.d/tunables/home.d"` instead (OR again we can either use the guard `if _, err := os.Stat("/etc/apparmor.d/tunables/home.d/"); err == nil`).

**Review by alexmurray — CHANGES_REQUESTED** [review]
I think we should either use `#include if exists` rather than just plain `#include` here otherwise we will fail to compile these profiles when running on systems where the kernel supports AppArmor but there is no apparmor userspace installed and hence there is no `/etc/apparmor.d/tunables/home.d/` file - OR we just add an extra guard like `if _, err := os.Stat("/etc/apparmor.d/tunables/home.d/"); err == nil` and keep using `#include`.

**Comment by alexmurray** [comment]
Thinking about this a bit more, in apparmor there is the file `/etc/apparmor.d/tunables/home` that `#include <tunables/home.d>` - so we should probably also manually include the `home` file as well in case the user has modified this instead of using the `tunables/home.d` directory - ie. something like the following:

```go
	snippet := ""
	tunables := []string{
		"/etc/apparmor.d/tunables/home",
                "/etc/apparmor.d/tunables/home.d",
	}
	for  _, f := range tunables {
		if _, err := os.Stat(f); err == nil {
			snippet += fmt.Sprintf(`#include "%s"` + "\n", f)
		}
	}
        return snippet
```

**Comment by Meulengracht** [comment]
> Thinking about this a bit more, in apparmor there is the file `/etc/apparmor.d/tunables/home` that `#include <tunables/home.d>` - so we should probably also manually include the `home` file as well in case the user has modified this instead of using the `tunables/home.d` directory - ie. something like the following:
> 
> ```go
> 	snippet := ""
> 	tunables := []string{
> 		"/etc/apparmor.d/tunables/home",
>                 "/etc/apparmor.d/tunables/home.d",
> 	}
> 	for  _, f := range tunables {
> 		if _, err := os.Stat(f); err == nil {
> 			snippet += fmt.Sprintf(`#include "%s"` + "\n", f)
> 		}
> 	}
>         return snippet
> ```

My only worry here (or just something to have in mind) - would to note that including <tunables/home> instead of <tunables/home.d> would effectively completely override any home tunables that are included in snapd (by resetting HOME/HOMEDIRS variables). Wouldn't correct protocol anyway for any user to use drop-ins in home.d and not home?

File: interfaces/apparmor/backend.go [hunk 4]
```diff
@@ -180,6 +180,19 @@ func snapConfineFromSnapProfile(info *snap.Info) (dir, glob string, content map[
...
+	// written to the system apparmor directory. However snapd vendors its own apparmor, which
+	// uses the readonly filesystem, which we cannot modify with our own snippets. So we force
+	// include the home.d tunables from /etc if necessary.
+	// XXX: Replace include home tunables until we have a better solution
+	features, _ := parserFeatures()
+	if strutil.ListContains(features, "snapd-internal") {
+		patchedProfileText = bytes.Replace(
+			patchedProfileText,
+			[]byte("#include <tunables/global>"),
+			[]byte("#include <tunables/global>\n#include \"/etc/apparmor.d/tunables/home.d/\""), -1)
```
[the remark this answers is not in the document]
Comment (Meulengracht) [reply interfaces/apparmor/backend.go]: good point! added `if exists`as that is more robust, should the presence of `home.d` ever change

File: interfaces/apparmor/backend.go (hunk 3 above)
Comment (Meulengracht) [reply interfaces/apparmor/backend.go]: same, fixed!

**Review by alexmurray — APPROVED** [review]
LGTM - it might be worth adding a comment that it is safe to use `#include if exists` in this context since the vendored apparmor should always support this (but perhaps we should have an additional check in the spread test to check this - something like the following just after we install the vendored apparmor snapd:

```shell
snap debug sandbox-features --required apparmor:parser:snapd-internal
snap debug sandbox-features --required apparmor:parser:include-if-exists
```

**Comment by alexmurray** [comment]
> My only worry here (or just something to have in mind) - would to note that including <tunables/home> instead of <tunables/home.d> would effectively completely override any home tunables that are included in snapd (by resetting HOME/HOMEDIRS variables). Wouldn't correct protocol anyway for any user to use drop-ins in home.d and not home?
> 
> If we followed this path, why not just include <tunables/global> from /etc/ and use those if present instead of the ones bundled with snapd? This would allow us to simply just replace <tunables/global> from a profile with #include "/etc/apparmor.d/tunables/global"

We can't just do `#include "/etc/apparmor.d/tunables/global"` as that has inside it `#include <tunables/home>` etc - and these include statements will get resolved to the `--base` path which is the vendored apparmor configuration - so it wouldn't work unfortunately.

For now we can probably just stick with `home.d` as is currently done, but I wouldn't be surprised if we may need to expand this to include either `home` or other tunables(.d) in the future.

**Comment by mvo5** [comment]
> > My only worry here (or just something to have in mind) - would to note that including <tunables/home> instead of <tunables/home.d> would effectively completely override any home tunables that are included in snapd (by resetting HOME/HOMEDIRS variables). Wouldn't correct protocol anyway for any user to use drop-ins in home.d and not home?
> > If we followed this path, why not just include <tunables/global> from /etc/ and use those if present instead of the ones bundled with snapd? This would allow us to simply just replace <tunables/global> from a profile with #include "/etc/apparmor.d/tunables/global"
> 
> We can't just do `#include "/etc/apparmor.d/tunables/global"` as that has inside it `#include <tunables/home>` etc - and these include statements will get resolved to the `--base` path which is the vendored apparmor configuration - so it wouldn't work unfortunately.
> 
> For now we can probably just stick with `home.d` as is currently done, but I wouldn't be surprised if we may need to expand this to include either `home` or other tunables(.d) in the future.

I wonder if apparmor itself should get an option in addition to base to be able to resolve tunables from a different location or having a earch path for tunables, wdyt?

**Comment by Meulengracht** [comment]
> I wonder if apparmor itself should get an option in addition to base to be able to resolve tunables from a different location or having a earch path for tunables, wdyt?

It'd be a neat feature to allow apparmor to resolve tunables from a different directory. Or just to have it possible to include `<tuneable>.d` folders from a different location.

**Review by mvo5 — APPROVED** [review]
Thank you!

File: interfaces/apparmor/backend.go [hunk 5]
```diff
@@ -829,6 +843,14 @@ func (b *Backend) addContent(securityTag string, snapInfo *snap.Info, cmdName st
 				return `#include if exists "/var/lib/snapd/apparmor/snap-tuning"`
 			}
 			return ""
+		// XXX: Remove this when we have a better solution to including the system
+		// tunables. See snapConfineFromSnapProfile() for a more detailed explanation.
+		case "###INCLUDE_SYSTEM_TUNABLES_HOME_D_WITH_VENDORED_APPARMOR###":
+			features, _ := parserFeatures()
+			if strutil.ListContains(features, "snapd-internal") {
+				return `#include if exists "/etc/apparmor.d/tunables/home.d"`
+			}
```
Comment (pedronis) [inline interfaces/apparmor/backend.go:852]: I don't see any unit test for this

**Review by pedronis — APPROVED** [review]
lgtm but there's a missing unit test

File: interfaces/apparmor/backend.go (hunk 5 above)
Comment (Meulengracht) [reply interfaces/apparmor/backend.go:852]: good point, added!

**Pull request merged**